File: iplan/db/operations/project.py
from typing import Mapping

from iplan.db.manager import connect_database
from iplan.db.models.project import Project

def create_project(name: str) -> Project:
    index = find_new_project_index()
    connection, cursor = connect_database()
    cursor.execute(f"INSERT INTO projects(name, i) VALUES ('{name}', {index})")
    connection.commit()
    return read_project(cursor.lastrowid)

def read_projects(archive: bool=False) -> Mapping[Project, list]:
    _filter = "WHERE archive = False"
    if archive:
        _filter = "ORDER BY archive ASC"
    query = f"SELECT * FROM projects {_filter}"

    connection, cursor = connect_database()
    records = cursor.execute(query).fetchall()
    return map(Project.new_from_record, records)

def read_project(project_id: int) -> Project:
    connection, cursor = connect_database()
    return Project.new_from_record(
        cursor.execute(
            f"SELECT * FROM projects WHERE id = {project_id}"
        ).fetchone()
    )

def update_project(project: Project, move_index=False) -> None:
    connection, cursor = connect_database()
    index_statement = ''

    if move_index:
        index_statement = f", i = {project.index}"
        old_project = read_project(project._id)

        projects_between = []
        step = 0
        range_condition = []    # start after or before old index to new index

        if old_project.index < project.index:
            range_condition = [">", "<="]
            step = -1
        elif old_project.index > project.index:
            range_condition = ["<", ">="]
            step = +1

        projects_between = cursor.execute(f"""SELECT * FROM projects WHERE
            i {range_condition[0]} {old_project.index} AND
            i {range_condition[1]} {project.index}
            """).fetchall()

        for record in projects_between:
            cursor.execute(
                f"""UPDATE projects SET
                i = {record[3]+step}
                WHERE id = {record[0]}"""
            )

    cursor.execute(
        f"""UPDATE projects SET
        name = '{project.name}',
        archive = {project.archive}
        {index_statement}
        WHERE id = {project._id}"""
    )
    connection.commit()

def delete_project(project: Project) -> None:
    connection, cursor = connect_database()
    cursor.execute(f"DELETE FROM projects WHERE id = {project._id}")
    cursor.execute(f"DELETE FROM lists WHERE project = {project._id}")
    cursor.execute(f"DELETE FROM tasks WHERE project = {project._id}")

    # decrease lower projects
    upper_projects = cursor.execute(
        f"SELECT * FROM projects WHERE i > {project.index}"
        ).fetchall()
    for record in upper_projects:
            cursor.execute(
                f"""UPDATE projects SET
                i = {record[3]-1}
                WHERE id = {record[0]}"""
            )

    connection.commit()

def search_projects(text: str, archive=False) -> Mapping[Project, list]:
    _filter = "AND archive = False"
    if archive:
        _filter = "AND ORDER BY archive ASC"
    query = f"SELECT * FROM projects WHERE name LIKE '%{text}%' {_filter}"

    connection, cursor = connect_database()
    records = cursor.execute(query).fetchall()
    return map(Project.new_from_record, records)

def find_new_project_index() -> int:
    connection, cursor = connect_database()
    last_index = cursor.execute("SELECT i FROM projects ORDER BY i DESC").fetchone()
    if not last_index:
        return 0
    return last_index[0] + 1

File: src/db/operations/project.rs
use std::cmp::Ordering;

use rusqlite::Result;

use crate::db::get_connection;
use crate::db::models::Project;

pub fn create_project(name: &str, icon: &str, description: &str) -> Result<Project> {
    let index = new_index();
    let conn = get_connection();
    conn.execute(
        "INSERT INTO projects(name, i, icon, description) VALUES (?1,?2,?3,?4)",
        (name, index, icon, description),
    )?;
    Ok(Project::new(
        conn.last_insert_rowid(),
        name.to_string(),
        false,
        index,
        icon.to_string(),
        description.to_string(),
    ))
}

pub fn read_projects(archive: bool) -> Result<Vec<Project>> {
    let filters = if !archive {
        "WHERE archive = false"
    } else {
        ""
    };
    let conn = get_connection();
    let mut stmt = conn.prepare(&format!("SELECT * FROM projects {filters} ORDER BY i ASC"))?;
    let mut rows = stmt.query([])?;
    let mut projects = Vec::new();
    while let Some(row) = rows.next()? {
        projects.push(Project::try_from(row)?)
    }
    Ok(projects)
}

pub fn read_project(project_id: i64) -> Result<Project> {
    let conn = get_connection();
    let mut stmt = conn.prepare("SELECT * FROM projects WHERE id = ?")?;
    stmt.query_row([project_id], |row| Project::try_from(row))
}

pub fn update_project(project: &Project) -> Result<()> {
    let conn = get_connection();
    let old_project = read_project(project.id())?;
    let index_stmt = &mut String::new();

    if project.index() != old_project.index() {
        index_stmt.push_str(&format!(", i = {}", project.index()));
        match project.index().cmp(&old_project.index()) {
            Ordering::Greater => {
                conn.execute(
                    "UPDATE projects SET i = i - 1
                    WHERE i > ?1 AND i <= ?2",
                    (old_project.index(), project.index()),
                )?;
            }
            Ordering::Less => {
                conn.execute(
                    "UPDATE projects SET i = i + 1
                    WHERE i < ?1 AND i >= ?2",
                    (old_project.index(), project.index()),
                )?;
            }
            Ordering::Equal => {}
        }
    }

    conn.execute(
        &format!(
            "UPDATE projects SET
            name = ?2, archive = ?3, icon = ?4, description = ?5 {index_stmt} WHERE id = ?1"
        ),
        (
            project.id(),
            project.name(),
            project.archive(),
            project.icon(),
            project.description(),
        ),
    )?;
    Ok(())
}

pub fn delete_project(project_id: i64, index: i32) -> Result<()> {
    let conn = get_connection();
    // Notify: Not return error when id not exists
    conn.execute("DELETE FROM projects WHERE id = ?", (project_id,))?;
    conn.execute("DELETE FROM sections WHERE project = ?", (project_id,))?;
    conn.execute("DELETE FROM tasks WHERE project = ?", (project_id,))?;
    // Decrease upper projects index
    conn.execute("UPDATE projects SET i = i - 1 WHERE i > ?1", (index,))?;
    Ok(())
}

pub fn find_projects(text: &str, archive: bool) -> Result<Vec<Project>> {
    let filters = if archive { "" } else { "AND archive = false" };
    // Replace % and _ with \% and \_ because they have meaning
    // FIXME: do this without copy string
    let text = text.replace('%', r"\%").replace('_', r"\_");
    let conn = get_connection();
    let mut stmt = conn.prepare(&format!(
        "SELECT * FROM projects WHERE name LIKE ? ESCAPE '\\' {filters}"
    ))?;
    let mut rows = stmt.query([format!("%{text}%")])?;
    let mut projects = Vec::new();
    while let Some(row) = rows.next()? {
        projects.push(Project::try_from(row)?)
    }
    Ok(projects)
}

fn new_index() -> i32 {
    let conn = get_connection();
    let mut stmt = conn
        .prepare("SELECT i FROM projects ORDER BY i DESC")
        .expect("Failed to find new index");
    let first_row = stmt.query_row([], |row| row.get::<_, i32>(0));
    match first_row {
        Ok(first_row) => first_row + 1,
        Err(_) => 0,
    }
}
